### DISCUSSION TITLE
Support for RTSP over HTTP

### DISCUSSION THREAD

**Discussion by Sanctus-dev** [dotnet/yarp#2648] [Q&A] [answered]
Hi, I'm looking for solutions to use YARP in contexts where supporting "RTSP over HTTP(S)" is a requirement.

Some short background to this protocol. Apple first published a [semi-official specification](https://www.yumpu.com/en/document/view/11877957/tunneling-quicktime-rtsp-and-rtp-over-http-relationship-to-) on how to traverse RTSP/RTP data [RFC2326](https://datatracker.ietf.org/doc/html/rfc2326#page-40) over firewalls using HTTP tunnels. The specification is now part of [ONVIF](https://www.onvif.org/), a specification used by many IP camera manufactures and Video Management Systems.

Basically the specification uses two long-lived GET and POST HTTP requests to stream media data and perform media control.

`The underlying RTSP over HTTP specification requires to open separate communication channels for continously sending and receiving data. For sending media from RTSP server to client the communication channel
is opened with a GET request and a response without content-length. For sending media to the server a POST
request is used. The specification notes that a content-length must be sent. Most applications use the value
32767 from the example given in the specification. Since http/2 streams may be closed by either side when
the content-length has been reached clients shall use a value large enough for the connection life time.`

I was not able to get this working with YARP. It seems like the GET request gets stuck in a response-body-copying phase. Typically, there will be NO content data on the GET channel until both the GET and POST channels have been established and a "DESCRIBE" command has been posted on the POST channel.

Is it a lost cause to try to get this playing?

**Comment by MihaZupan** [comment]
If the traffic is valid HTTP, I don't see why it couldn't work.
If you have multiple destinations behind the proxy, you might have to ensure that load balancing doesn't distribute the two requests to different backend servers.
You may also have to check that the right HTTP version is being used if all peers expect a specific version.

The first spec you linked mentions
> "All POST requests are required to have a content-length header by HTTP. In practice the actual value seems to be ignored by proxy servers, so it's possible to send more than this amount of data"

This is just wrong, doing so is a protocol violation, and won't work with ASP.NET/YARP. You should be able to check if something like that is happening by looking at a network trace for example.

**Reply by Tratcher** [reply]
Yeah, better to use HTTP/1.1 Transfer-Encoding: Chunked or HTTP/2 with no content-length (framed similar to chunking) in both directions.

The response stall might just be a matter of needing to flush the headers. That's probably doable in a response transform. YARP does this automatically for gRCP and WebSockets, but not for normal GET requests.

**Reply by Sanctus-dev** [reply]
The problem is happening when establishing the GET-tunnel before the POST channel is even created. YARP seems to be waiting and then doing a timeout while waiting on the content of the GET response body (which will not return any data at this stage).  Here is the call-stack:

`Yarp.ReverseProxy.Forwarder.HttpForwarder: Warning: ResponseBodyCanceled: **Copying the response body was canceled**.
System.Threading.Tasks.TaskCanceledException: The operation was canceled.
 ---> System.IO.IOException: Unable to read data from the transport connection: The I/O operation has been aborted because of either a thread exit or an application request..
 ---> System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Int32>.GetResult(Int16 token)
   at System.Net.Http.HttpConnection.ReadAsync(Memory1 destination)
   at System.Net.Http.HttpConnection.ConnectionCloseReadStream.ReadAsync(Memory1 buffer, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.ConnectionCloseReadStream.ReadAsync(Memory1 buffer, CancellationToken cancellationToken)
   at Yarp.ReverseProxy.Forwarder.StreamCopier.CopyAsync(Stream input, Stream output, Int64 promisedContentLength, StreamCopierTelemetry telemetry, ActivityCancellationTokenSource activityToken, Boolean autoFlush, CancellationToken cancellation)
Yarp.ReverseProxy.Forwarder.HttpForwarder: Information: Proxying to http://172.25.68.62/axis-cgi/com/ptzqueue.cgi?control=query&camera=1 HTTP/2 RequestVersionOrLower 
Yarp.ReverseProxy.Forwarder.HttpForwarder: Information: Received HTTP/1.1 response 200.`

I have looked at and compared wiresharks of this and a direct-working connection, and my conclusion is the same as above. I can post wiresharks here if needed.

**Reply by Sanctus-dev** [reply]
> Yeah, better to use HTTP/1.1 Transfer-Encoding: Chunked or HTTP/2 with no content-length (framed similar to chunking) in both directions.
> 
> The response stall might just be a matter of needing to flush the headers. That's probably doable in a response transform. YARP does this automatically for gRCP and WebSockets, but not for normal GET requests.

@Tratcher - Unfortunately I have no control of the clients in this use-case, but good to know. Flushing the headers sound interesting - could you point me to some resource explaining/exemplifying this?

**Reply by Tratcher** [reply]
Here's a sample on adding response transforms:
https://github.com/microsoft/reverse-proxy/blob/f1c90a1fe8987a527b25d22bf6850257228c15b4/testassets/ReverseProxy.Code/Program.cs#L86
Something like this should do:
`context.HttpContext.Response.Body.FlushAsync()`

**Reply by MihaZupan** [reply]
If that doesn't work, please do share the network captures (including the working version without the proxy).

**Reply by MihaZupan** [reply]
Were you able to get this working?

**Reply by Sanctus-dev** [reply]
So I've been fiddling around with this for a while now and I managed to progress a little bit - much thanks to the Flush call. The client now manages to to initiate the two requests (GET/POST) and send the initial request but for some reason the server never post a reply on the GET channel. Eventually the client will timeout.

I'm trying to compare network-traces and understand what differs, One obvious difference is that the proxy enables chunked transfer encoding which makes the request payload not match exactly - I'm unsure if the server supports this. Is it possible to turn it off?

**Reply by MihaZupan** [reply]
If you suspect chunking may be a problem, you could try lying to the server and set a huge `Content-Length` header (e.g. 1 TB).
This will mean every request will involve an exception at some point (we never reach the end), but it may work if the server is (erroneously) expecting a raw data stream.

**Reply by Sanctus-dev** [reply]
@MihaZupan - I mostly tested by _removing_ the content-length header on the POST request - that has been the most promising approach. If I keep or change the content-length I get errors from the proxy, e.g.
"System.InvalidOperationException: Sent 226 request content bytes, but Content-Length promised 32767"

To summarize my approach so far:

For the GET request:

- AddResponseTransform
- responseContext.HttpContext.Response.Body.FlushAsync()

For POST request:

- AddRequestTransform
- Create new MemoryStream
- Peak and copy the request body stream to the MemoryStream until end of the first RTSP-request is found (see details below)
- Flush the StreamWriter
- Set MemoryStream.Position to 0
- Set requestContext.HttpContext.Request.Body = MemoryStream

Determining the end of the a RTSP request is a bit tricky since the request-data is base64-encoded. You need to continuously decode the stream looking for double line-breaks ("\r\n\r\n").

My first goal is to get the response of the first RTSP-request back on the GET-channel.
The next goal would be to get this to work with multiple RTSP-requests.

**Reply by MihaZupan** [reply, marked as answer]
Any luck with collecting network traces with & without the proxy?

You only have two ways of sending the request content - `Content-Length`, or `chunked`.
If you suspect chunking is the problem, lying about the content length is the only option. The `"Sent X request content bytes, but Content-Length promised Y"` are the errors I was referring to that you would always see if you do so, but they may be harmless in your case.

**Comment by galvesribeiro** [comment]
I've been looking at this myself. I have a DVR which support ONVIF and exports all 32 chambers (coax) connected on it on the ONVIF response. However, the ONVIF client that I have (Ubiquiti Unifi Protect NVR) expect that each camera to have its own IP (yeah they are wrong and not 100% implementing ONVIF). 

So, what I was thinking of, is to have an application with YARP which would listen to multiple hosts (DNS, IP, URL path, whatever) and redirect It to a mapped channel on this single DVR, proxying and formatting the parameters in order to make this proxying. 

All we care right now is the camera feed. Will keep an eye here and report back what I've found while experimenting with YARP in this scenario.
